File: Main-Project/StudySpace/src/main/java/application/studyspace/services/API/DeepSeekAPI.java
package application.studyspace.services.API;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import org.json.JSONObject;

/**
 * The DeepSeekAPI class provides a method to interact with the DeepSeek AI model
 * by making HTTP requests to the specified API endpoint, sending user prompts,
 * and retrieving AI-generated responses.
 *
 * This class encapsulates the API communication logic, including payload construction,
 * HTTP request handling, and response parsing. It supports sending dynamic input
 * directly to the AI model and handles any errors that may occur during the process.
 */
public class DeepSeekAPI {

    private static final String API_URL = "https://api.fireworks.ai/inference/v1/chat/completions";
    private static final String API_KEY = "fw_3ZWJHuiuLsE4Sh1zNXhq751E";

    /**
     * Executes a DeepSeek API call with dynamic input and returns the AI's message response.
     *
     * @param dynamicInput The input question or prompt to send to the API.
     * @return The AI's response message as a string.
     */
    public static String executeDeepSeekAPI(String dynamicInput) {
        try {
            JSONObject payload = new JSONObject();
            payload.put("model", "accounts/fireworks/models/deepseek-v3");
            payload.put("max_tokens", 16384);
            payload.put("top_p", 1);
            payload.put("top_k", 40);
            payload.put("presence_penalty", 0);
            payload.put("frequency_penalty", 0);
            payload.put("temperature", 0.6);

            JSONObject message = new JSONObject();
            message.put("role", "user");
            message.put("content", dynamicInput);

            payload.put("messages", new JSONObject[] { message });

            // Build the HTTP request
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(API_URL))
                    .header("Accept", "application/json")
                    .header("Content-Type", "application/json")
                    .header("Authorization", "Bearer " + API_KEY) // Add your API key
                    .POST(HttpRequest.BodyPublishers.ofString(payload.toString()))
                    .build();

            // Send the request and handle the response
            HttpClient client = HttpClient.newHttpClient();
            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

            // Parse the response to extract the AI's message
            JSONObject jsonResponse = new JSONObject(response.body());
            String aiResponse = jsonResponse.getJSONArray("choices")
                    .getJSONObject(0)
                    .getJSONObject("message")
                    .getString("content");

            return aiResponse;

        } catch (Exception e) {
            return "Error occurred during DeepSeek API call: " + e.getMessage();
        }
    }

    public static class ExampleUsage {
        public static void main(String[] args) {
            String input = "Hello, can you summarize the theory of relativity?";
            String response = DeepSeekAPI.executeDeepSeekAPI(input);
            System.out.println("DeepSeek Response: " + response);
        }
    }

}

File: Main-Project/StudySpace/src/main/java/application/studyspace/services/API/DeepSeekAPI.js
import { InferenceClient } from "@huggingface/inference";

/**
 * Represents an instance of the `InferenceClient` used to interact with an external API.
 *
 * The client is initialized with an API key retrieved from environment variables.
 * This key is used to authenticate with the API and access its services.
 *
 * @type {InferenceClient}
 */
const client = new InferenceClient(process.env.API_KEY);

/**
 * Sends a user-provided input to the DeepSeek API and retrieves a response from the AI model.
 *
 * @param {string} userContent - The content or message provided by the user to be processed by the AI model.
 * @return {Promise<string>} A promise that resolves to the AI's response as a string. If no response is received, returns "No response from AI". If an error occurs, logs the error and does not return a value.
 */
async function DeepSeekAPI(userContent) {
  try {
    const chatCompletion = await client.chatCompletion({
      provider: "fireworks-ai",
      model: "deepseek-ai/DeepSeek-V3-0324",
      messages: [
        {
          role: "user",
          content: userContent,
        },
      ],
    });


    const response = chatCompletion.choices[0]?.message?.content || "No response from AI";
    console.log(response); // This is sent to stdout
    return response;
  } catch (error) {

    const errorMsg = "Error querying the Hugging Face API: " + (error.message || error);
    console.error(errorMsg);
    console.log(errorMsg);
  }
}
DeepSeekAPI();
